### PR TITLE
chore: update k8s libs to 1.24

### PR DESCRIPTION
**Pull request by mengqiy**
We want to bump the k8s dependencies, since when want to import a newer kyaml version in kpt. This kyaml version is only compatible with k8s v0.24+ due import path change in https://github.com/kubernetes/kube-openapi/pull/285.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -7,20 +7,20 @@ require (
...
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
-	k8s.io/api v0.23.2
-	k8s.io/apimachinery v0.23.2
-	k8s.io/cli-runtime v0.23.2
-	k8s.io/client-go v0.23.2
-	k8s.io/component-base v0.23.2
-	k8s.io/klog/v2 v2.30.0
-	k8s.io/kubectl v0.23.2
-	k8s.io/utils v0.0.0-20211208161948-7d6a63dca704
+	k8s.io/api v0.24.0-beta.0
```
Comment (mengqiy) [inline go.mod:14]: v0.24.0-rc.0 is expected today.
I will update once it's out.

File: cmd/diff/cmddiff.go [hunk 2]
```diff
@@ -80,17 +80,12 @@ func Initialize(o *diff.DiffOptions, f util.Factory, args []string) (func(), err
...
-		return cleanupFunc, err
-	}
-
 	o.DynamicClient, err = f.DynamicClient()
 	if err != nil {
 		return cleanupFunc, err
 	}
 
-	o.DryRunVerifier = resource.NewDryRunVerifier(o.DynamicClient, o.DiscoveryClient)
+	o.DryRunVerifier = resource.NewQueryParamVerifier(o.DynamicClient, f.OpenAPIGetter(), resource.QueryParamDryRun)
```
Comment (mengqiy) [inline cmd/diff/cmddiff.go:88]: This change is similar to https://github.com/kubernetes/kubectl/commit/7e681490ec30a64626b359b7346d891240155794

File: pkg/kstatus/polling/clusterreader/caching_reader.go [hunk 3]
```diff
@@ -292,8 +292,8 @@ func (c *CachingClusterReader) listUnstructured(
 		if t.ResourceVersion != "" {
 			u.SetResourceVersion(t.ResourceVersion)
 		}
-		if t.SelfLink != "" {
-			u.SetSelfLink(t.SelfLink)
+		if t.SelfLink != "" { // nolint:staticcheck
```
Comment (mengqiy) [inline pkg/kstatus/polling/clusterreader/caching_reader.go:295]: > SA1019: t.SelfLink is deprecated: selfLink is a legacy read-only field that is no longer populated by the system. +optional (staticcheck)

This field is deprecated. Skipping it.

**Comment by yuwenma** [comment]
This PR will fix the go mod conflicts we had in kpt and porch, and it is critical for all the users who use both KRM functions (new sdk) and cli-utils!

**Comment by karlkfi** [comment]
Historically we haven't taken new k8s versions until they're fully released and can be picked up by the known downstream clients (kpt, Config Sync, GitLab Agent). I'm not sure how to handle a release candidate. My best advice would be to temporarily vendor cli-utils with your changes into your downstream client, until cli-util is updated with the change and a new release is cut.

**Comment by mengqiy** [comment]
Rebased. PTAL

**Comment by karlkfi** [comment]
/hold

Blocking until https://github.com/kubernetes-sigs/cli-utils/pull/572 is merged.

**Review by karlkfi — APPROVED** [review]
/lgtm

**Review by mortent — APPROVED** [review]
/approve

**Pull request merged**